testng-core/src/main/java/org/testng/internal/invokers/BaseInvoker.java

```diff
@@ -54,19 +57,40 @@ protected void runInvokedMethodListeners(
 
     InvokedMethodListenerInvoker invoker =
         new InvokedMethodListenerInvoker(listenerMethod, testResult, testResult.getTestContext());
-    for (IInvokedMethodListener currentListener : m_invokedMethodListeners) {
-      try {
-        invoker.invokeListener(currentListener, invokedMethod);
-      } catch (SkipException e) {
-        String msg =
-            String.format(
-                "Caught a [%s] exception from one of listeners %s. Will mark [%s()] as SKIPPED.",
-                SkipException.class.getSimpleName(),
-                currentListener.getClass().getName(),
-                invokedMethod.getTestMethod().getQualifiedName());
-        Utils.warn(msg);
-        testResult.setStatus(ITestResult.SKIP);
-        testResult.setThrowable(e);
+    if (InvokedMethodListenerMethod.BEFORE_INVOCATION.equals(listenerMethod)) {
```

Comment:
The code is being duplicated in both the cases. Instead we should just resort to invoking `Collections.reverse()` if `listenerMethod != BEFORE_INVOCATION`

Also when it comes to enums, we should be using `==` (or) `!=` for comparison instead of `equals()`

Reply:
Do you mean the `try{}catch()` part for `listenerMethod == BEFORE_INVOCATION` is not needed?

Reply:
No. I am not saying that. I am saying that we should avoid code duplication here. The only thing that is changing in the if and else part is the reversing. So can we have that commonly outside and then have the rest of it done only once ?

Reply:
The `catch(){}` part prints `currentListener.getClass().getName()`, the `currentListener` is valid in each `for(){}`, but the two cases has its own `for(){}`. Do you have any ideas how to avoid duplicate?

Reply:
I think I got you what you said, but it changes the listener default order, do we need copy the listener to keep the state?
```
 if (InvokedMethodListenerMethod.BEFORE_INVOCATION != listenerMethod) {
    	Collections.reverse((List<IInvokedMethodListener>) m_invokedMethodListeners);
    }
      for (IInvokedMethodListener currentListener : m_invokedMethodListeners) {
        try {
          invoker.invokeListener(currentListener, invokedMethod);
        } catch (SkipException e) {
          String msg =
              String.format(
                  "Caught a [%s] exception from one of listeners %s. Will mark [%s()] as SKIPPED.",
                  SkipException.class.getSimpleName(),
                  currentListener.getClass().getName(),
                  invokedMethod.getTestMethod().getQualifiedName());
          Utils.warn(msg);
          testResult.setStatus(ITestResult.SKIP);
          testResult.setThrowable(e);
        }
      }
  }
```

Reply:
I am suggesting something like this

```java
    boolean isBeforeInvocation = InvokedMethodListenerMethod.BEFORE_INVOCATION == listenerMethod;
    Collection<IInvokedMethodListener> listeners = isBeforeInvocation ? m_invokedMethodListeners :
        Lists.newReversedArrayList(m_invokedMethodListeners);
    for (IInvokedMethodListener currentListener : listeners) {
      try {
        invoker.invokeListener(currentListener, invokedMethod);
      } catch (SkipException e) {
        String msg =
            String.format(
                "Caught a [%s] exception from one of listeners %s. Will mark [%s()] as SKIPPED.",
                SkipException.class.getSimpleName(),
                currentListener.getClass().getName(),
                invokedMethod.getTestMethod().getQualifiedName());
        Utils.warn(msg);
        testResult.setStatus(ITestResult.SKIP);
        testResult.setThrowable(e);
      }
    }
```

We also would need to add anew method like below into `org.testng.collections.Lists`

```java
  public static <K> List<K> newReversedArrayList(Collection<K> c) {
    List<K> list = newArrayList(c);
    Collections.reverse(list);
    return list;
  }
```